PowerShell Bits, tutorial series (#7278)
MicrosoftDocs/PowerShell-Docs: reference/docs-conceptual/learn/tutorials/01-discover-powershell.md

## Cmdlets, compiled commands

Upon install, you get a lot of commands installed. Commands, in PowerShell, are called cmdlets, it's pronounced as "command-let".

Every cmdlet is built up of a verb and a noun. This convention makes it easier to understand what it does. Additionally, that fact also make it easier to find the command your looking for, as you can filter on verb and noun, when looking for a specific cmdlet to use. A typical cmdlet therefore has the following naming convention "Verb-Noun", with an example being `Get-Process`.

### Core cmdlets

When you first pick up PowerShell, it might feel intimidating as there seem to be so much to learn. However, PowerShell is meant to be learned little at a time. Having that mindset going forward, lets you take one problem at a time.

Something that will help you on your journey, are the so called _core cmdlets_, commands that helps you discover PowerShell. Using the mentioned _core cmdlets_ you can discover what commands there are, what they do, and what types they operate on. Let's list the core cmdlets below:

- `Get-Command`. This command retrieves a list of all commands installed on your machine.

> - `Get-Help`. By invoking this command, with the name of a command as an argument, you'll be shown a help page describing various parts of a command.

Author: @sdwheeler  it's in the next tutorial i.e using help, updating it etc, such a big topic

Cmdlet          Stop-Process                                       7.0.0.0    Microsoft.PowerShell.Management
Cmdlet          Wait-Process                                       7.0.0.0    Microsoft.PowerShell.Management
```

### Helper cmdlets

**Take n records**
-Last
-First <n records>
Get-Command | Select-Object -first 3

**select objects based on prop value**
Where-Object {expression}
Get-Process | Where-Object {$_.ProcessName -Match "^p.*"}

**Sorting**
Sort-Object -Descending // default sorts on name
Sort-Object -Property <property>
Get-Command | Select-Object -first 3 | Sort-Object -Descending

**Group by**
group by 

## Explore a command with Get-Member 

another way to approach your learning PS is to find related commands. related commands can be defined as commands operating on the same type. So how do we know that = get-member, gives us the type. once we have the type, we can use that as an argument when we use get-command

Get-Process | Get-Member # Process
Get-Process -ParameterType Process # shows all commands operating on that type

## Exercise - calling your first command

## Exercise 

> Select a command we want to know more about
> Inspect command and find type
> Find related commands

Reviewer: ```suggestion

1. Select a command we want to know more about
2. Inspect command and find type
3. Find related commands

```

```

### Helper cmdlets, that helps with filtering

Apart from using parameters to filter, you can also use commands to help you with this task. Here's some commands that can act as filters:

- **Select-Object**. It's a very versatile command that helps you pick out specific properties from one or more objects. Additionally by using it's parameters you can limit the response you get back. Here's an example of `Select-Object` being used to ask for a limited number of records:

   ```powershell
   Get-Command | Select-Object -First 3
   ```

   The result from the above is the three first commands, counted from the top. The result looks like so:

   ```outline
   CommandType     Name                                               Version    Source
   -----------     ----                                               -------    ------
   Alias           Add-AdlAnalyticsDataSource                         1.0.2      Az.DataLakeAnalytics
   Alias           Add-AdlAnalyticsFirewallRule                       1.0.2      Az.DataLakeAnalytics
   Alias           Add-AdlStoreFirewallRule                           1.3.0      Az.DataLakeStore
   ```

>    It's worth looking into to this command further as it can do a lot more [docs Select-Object](https://docs.microsoft.com/en-us/powershell/module/microsoft.powershell.utility/select-object?view=powershell-7.1)

Author: ```suggestion
   It's worth looking into to this command further as it can do a lot more [docs Select-Object](https://docs.microsoft.com/powershell/module/microsoft.powershell.utility/select-object?view=powershell-7.1&preserve-view=true)
```

    Cmdlet          Enter-PSHostProcess                                7.1.0.0    Microsoft.PowerShell.Core
    Cmdlet          Get-Process                                        7.0.0.0    Microsoft.PowerShell.Managem…
    Cmdlet          Get-PSHostProcessInfo                              7.1.0.0    Microsoft.PowerShell.Core
    Cmdlet          Stop-Process                                       7.0.0.0    Microsoft.PowerShell.Managem…
    Cmdlet          Wait-Process                                       7.0.0.0    Microsoft.PowerShell.Managem…
   ```

   Congrats, you've managed to find other commands that operates on the same type `Process`. Using `Get-Member`, is a good starting point to understand what other commands you should check out next.

## Summary

In this first part, you learned what PowerShell is and what areas in can be used. You where then taught about cmdlets and in particular the so called _core cmdlets_. Knowing the core cmdlets is important as it teaches you how to learn. Out of the the three core cmdlets, you were taught to use `Get-Command` and `Get-Member`. In the next part, you'll learn how to use the powerful help system.

> 

Author: ###
```suggestion

### Additional resources

- [reference doc, Get-Command](https://docs.microsoft.com/powershell/module/microsoft.powershell.core/get-command?view=powershell-7.1&WT.mc_id=academic-0000-chnoring)
- [reference doc, Get-Member](https://docs.microsoft.com/powershell/module/microsoft.powershell.utility/get-member?view=powershell-7.1&preserve-view=true&WT.mc_id=academic-0000-chnoring)
```

Cmdlet          Get-Process                                        7.0.0.0    Microsoft.PowerShell.Management
Cmdlet          Start-Process                                      7.0.0.0    Microsoft.PowerShell.Management
Cmdlet          Stop-Process                                       7.0.0.0    Microsoft.PowerShell.Management
Cmdlet          Wait-Process                                       7.0.0.0    Microsoft.PowerShell.Management
```

### Helper cmdlets, that helps with filtering

Apart from using parameters to filter, you can also use commands to help you with this task. Here's some commands that can act as filters:

- **Select-Object**. It's a very versatile command that helps you pick out specific properties from one or more objects. Additionally by using it's parameters you can limit the response you get back. Here's an example of `Select-Object` being used to ask for a limited number of records:

   ```powershell
   Get-Command | Select-Object -First 3
   ```

   The result from the above is the three first commands, counted from the top. The result looks like so:

>    ```outline

Author: ```suggestion
   ```output
```

Cmdlet          Get-Process                                        7.0.0.0    Microsoft.PowerShell.Managem…
Cmdlet          Get-PSHostProcessInfo                              7.1.0.0    Microsoft.PowerShell.Core
Cmdlet          Stop-Process                                       7.0.0.0    Microsoft.PowerShell.Managem…
Cmdlet          Wait-Process                                       7.0.0.0    Microsoft.PowerShell.Managem…
```

As you can see, knowing the type of a command, can greatly narrow down your search after commands that might be interesting for you.

## Exercise - calling your first command

In this exercise, you'll learn how to run your first command.

1. Start a PowerShell console by typing `pwsh`:

   ```powershell
   pwsh
   ```

1. Run the following `$PSVersionTable.PSVersion`:

   ```powershell
   $PSVersionTable.PSVersion
   ```

   Your output looks similar to this:

   ```output
   Major  Minor  Patch  PreReleaseLabel BuildLabel
   -----  -----  -----  --------------- ----------
   7      1      0 
   ```

   Congrats, you've successfully run your first command and was able to get information of what version of PowerShell is installed on your system.

> ## Exercise - find related commands

Author: I alreeady did Get-Process in the concept part. Thinking this should be something else. Ideas @sdwheeler ?

   It's worth looking into to this command further as it can do a lot more
   [docs Select-Object](/powershell/module/microsoft.powershell.utility/select-object)

- `Where-Object`. The where object helps you select objects from a collection based on the values
  of properties. The command takes an expression in which you are able to express what column/s you
  want to match against what values. To find all process object where the `ProcessName` starts with
  `p`, you could use `Where-Object` like so:

  ```powershell
  Get-Process | Where-Object {$_.ProcessName -Like "p*"}
  ```

  Above, the `Get-Process` cmdlet produces a collection of process objects. To filter down the
  response, you _pipe_ the command `Where-Object`. Piping means that two or more commands are
  connected via a pipe `|` character. The idea is that the output from one command serves as the
  input for the next command, as read from left to right. The `Where-Object` uses an expression to
  filter. The expression itself uses the `-Like` operator and string argument that is a wildcard
  expression.

> ## Explore a objects with Get-Member

Author: ```suggestion
## Explore objects with Get-Member
```
